### ISSUE TITLE
go/ast: free-floating comments are single-biggest issue when manipulating the AST

### ISSUE THREAD

**Issue by griesemer** [golang/go#20744]
Reminder issue.

The original design of the go/ast makes it very difficult to update the AST and retain correct comment placement for printing. The culprit is that the comments are "free-floating" and printed based on their positions relative to the AST nodes. (The go/ast package is one of the earliest Go packages in existence, and this is clearly a major design flaw.)

Instead, a newer/updated ast should associate all comments (not just Doc and Comment comments) with nodes. That would make it much easier to manipulate a syntax tree and have comments correctly attached. The main complexity with that approach is the heuristic that decides which comments are attached to which nodes.

**Comment by jimmyfrasche** [comment]
I've made (unattached) comments a node type in various toy/simple
parsers where I needed to preserve comments and it worked well for my
needs

**Comment by griesemer** [comment]
@jimmyfrasche Thanks for your comment. I do have a design and prototype that is close, but not good enough for gofmt purposes. I'd be interested in how you integrate your comment nodes in the rest of a syntax tree.

**Comment by jimmyfrasche** [comment]
To put it in terms of the existing go/ast:

All attached comments would remain where they are on various nodes.

Comment would be both a Decl and a Stmt.

Unattached Comments would be interspersed with declarations and
statements where they appeared in the source.

For go/ast you'd also need a second CommentGroup for comments before
the comments attached to the package declaration.

If you don't parse comments, they aren't included.

If you do, sometimes you have to skip comment nodes when walking the
AST, but you can add and delete comments very easily and preserve
comments when modifying the AST just as easily.

I've done this one or twice with much simpler parsers/formatters built
for much simpler purposes, but it worked well for my uses.

**Comment by jimmyfrasche** [comment]
I should add that when I did that they were languages that only had
"until end of line" comments so there would still be complications in
cases like fmt.Println(x /* + y */)

**Comment by griesemer** [comment]
@jimmyfrasche Thanks for the clarification. Because /* */ style comment can appear anywhere, they do add complexity as it's not always clear if they belong to the token before or after.

**Comment by jimmyfrasche** [comment]
Everything stores a position so it could be worked out, though that
makes it hard to preserve the comments when updating code with inline
comments.

**Comment by jimmyfrasche** [comment]
I suppose one way around that would be, in expressions, to always have comments come after (where, for binops "after" is defined as after the binary operator so that
```
a /* 1 */ + /* 2 */ b /* 3 */
```
stores 1 on a, 2 on +, and 3 on b.

Stuff like CallExpr would need to store multiple comments for code like
```
f /* 1 */ () /* 2 */
```
and statements would need before and after comments for
```
/* before */ return /* after - but if expression is returned comment is attached to that */
```

With that and the "comments are nodes" approach plus the extra CommentGroup on the File, I think that would get everything.

Verbose and filled with cases but explicit and easy to manipulate.

**Comment by griesemer** [comment]
@jimmyfrasche My approach is similar. On the one extreme, each token could have a potential comment attached; and on the other, there's one comment for each node. The former is too expensive in terms of representation but would probably preserve comment positioning perfectly; and the latter is relatively cheap but may not preserve all comments in place (depending on heuristics). The stored position is helpful, but relying on it introduces the same problems we have already. Thus, any solution will have to work based on the graph alone, and ignore positions (they are used only in the beginning for the heuristics to determine how to associate a node).

**Comment by quasilyte** [comment]
Are there any advices in how this can be done today?

In my experience, trailing comments are particularly hard to handle. 
Suppose a tool that sorts constants in alphabetical order:
```go
const (
  C // C comment
  A
  B // B comment
)
```
With manual `Pos` update and `format.Node`, the best result I managed to achieve is:
```go
const (
  A
  // C comment
  C 
  // B comment
  B
)
```

I am sorry if this is just me being lame.

#17108 had some progress, but it is unclear if it has any relation to this.

**Comment by ianlancetaylor** [comment]
@Quasilyte This issue is about fixing the problem.  Let's not complicate the issue with a discussion of how to deal with the problem before it is fixed.  Please take that discussion to a forum; see https://golang.org/wiki/Questions.  Thanks.

**Comment by matthewmueller** [comment]
Hey guys, I'm wondering if I can help push this one along in some way. I keep coming up with reasons to modify, append and delete pieces of the AST, but positioning comments ends up biting pretty hard.

The way I was thinking about this is rather similar to https://github.com/golang/go/issues/20744#issuecomment-310188325. 

I'm not 100% sure how this would work at the token / parsing level, but I think the AST changes would have the following characteristics:

- comment nodes implement `ast.Decl` and `ast.Stmt`
- free-floating comments (one or more spaces below) would be scattered throughout the AST, in the same way declarations and statements are.
- attached comments (on top of or to the side of a field), would be considered owned by the node and populate the `Doc` field.

And then, any changes to these comment nodes would be reflected when `ast/printer` traverses the tree and prints the code out.

**Comment by griesemer** [comment]
@matthewmueller This sounds about right to me, but the devil is in the details.

I think we would want each AST node to have an additional comment field (that is usually a nil pointer) and that field would then be used to attach comments that "belong" (are associated) with that node. There may be additional information about the comment's position needed (as in "above", "before", "after", or "below" the node's line) for good positioning. Such associated comments would be always attached with the relevant nodes. (It's not clear to me that we can just use the existing Doc fields and expand the concept of Doc fields to all nodes given that those have specific meaning at the moment.) Additionally, as you have observed, "free-floating" (non-associated) comments would need to act like Decls and Stmts (and presumably those would always have an empty line before and after them as otherwise they would be associated with a node.

It should be possible to do this is an backward-compatible way: Add a new *Comment (or similar) field to all nodes, and define the respective data structure. Then, a new function could take the existing list of comments from the parser and associate them with the nodes. This is where a good heuristic is needed (I've had a prototype for this a couple of years back but I didn't get to make it complete). Finally, go/printer would need to be adjusted to position comments based on node association (I'd copy go/printer and create a 2nd version that is appropriately modified - this would make sure the existing code will continue to run as always).

The hard part is getting the comment association heuristic work well.

**Comment by josharian** [comment]
@matthewmueller some concrete examples of cases that are currently mishandled: #21755, #22371.

**Comment by dave** [comment]
Hey all, I'm currently working on a project that attempts to address this problem.

[github.com/dave/dst](https://github.com/dave/dst) is a fork of the `go/ast` package with a few changes. The [decorator](https://github.com/dave/dst/tree/master/decorator) package converts from `*ast.File + *token.FileSet` to `*dst.File` and back again.

All the position fields have been removed from `dst` so it's just the location in the tree that determines the position of the tokens. Decorations (e.g. comments and newlines) are stored along with each node, and attached to the node at various points. The intention is that any place `gofmt` will allow a comment / new-line to be attached, `dst` will allow this.

I've finished a very rough prototype that works pretty well. (Take a look at [restorer_test.go](https://github.com/dave/dst/blob/master/decorator/restorer_test.go#L11) - all the tests pass apart from `FuncDecl` now).

There's several special cases that it doesn't currently handle. Right now I'm generating much of the code, so the special cases are non-trivial to implement. (e.g. Look at [FuncDecl](https://github.com/golang/go/blob/master/src/go/ast/ast.go#L927-L934) - the `func` token from the `Type` field is rendered before `Recv` and `Name`). Over the next few weeks I'll refactor and handle the special cases.

As @griesemer points out a big problem is where to attach the decorations so as you manipulate the tree they remain attached to the node you were expecting. My algorithm probably needs improvement here too (see [decorator_test.go](https://github.com/dave/dst/blob/master/decorator/decorator_test.go)), but I think it currently works well enough to be useful.

This issue is concerned with a replacement for the `go/ast` package - `dst` isn't necessarily aiming to be this, so I'll invite further conversation in the [dst repo issues](https://github.com/dave/dst/issues) or in the [#dst](https://gophers.slack.com/messages/CCVL24MTQ) Gophers Slack channel.

**Comment by dave** [comment]
I'm pretty happy with how development of github.com/dave/dst has come on in the past few weeks. 

My main yardstick is [a test](https://github.com/dave/dst/blob/master/decorator/restorer_std_test.go) that converts the entire Go standard library from `ast` to `dst` and back to `ast`. It prints the resultant `ast` and compares with the original source. Using this I've found a host of edge cases which are now all fixed. It's possible there's more bugs but I think the standard library is a pretty good corpus of code, and it's now working with 100% fidelity.

I'm also a lot happier with the algorithm that attaches comments to nodes. It gives acceptable results in the vast majority of cases.

Over the next few weeks I'll be dog-fooding the system in a couple of projects, so I'll get a better idea of how it works in the real-world. 

I would invite you all to take a look. I may make some tweaks to the API that controls adding / removing comments from the attachment points, and I'd very much appreciate feedback on this.

**Comment by stapelberg** [comment]
> I would invite you all to take a look. I may make some tweaks to the API that controls adding / removing comments from the attachment points, and I'd very much appreciate feedback on this.

I have just gotten a chance to test the `dst` package, and it indeed deals with newlines/whitespace much better than the `go/ast` package in that `dst` allows you to insert line breaks/blank lines before/after inserted AST nodes.

I haven’t had the need to touch comments for my use-case (yet?), but the whitespace controls are very valuable, and I did not run into any correctness issues with `dst` yet.

Thanks for the helpful package!

**Comment by ZhengHe-MD** [comment]
@dave the dst package just saved my life. it's very helpful!

**Comment by Zemnmez** [comment]
@dave really appreciate dst. this was such an issue for generating good code

**Comment by mdempsky** [comment]
>  On the one extreme, each token could have a potential comment attached; and on the other, there's one comment for each node. The former is too expensive in terms of representation but would probably preserve comment positioning perfectly; and the latter is relatively cheap but may not preserve all comments in place (depending on heuristics).

I've been brainstorming lately about generalizing unified IR to a general serialization format for Go syntax+types and its possible benefits. (Ideally, that we only need to implement a single parser+typechecker that directly writes out a binary serialization; and then multiple, much simpler reader implementations that construct from that common serialization format either compiler IR, go/ast+go/types IR, or some hypothetical go2/* IR.)

In that world, I think it's practical to imagine keeping full-fidelity position information for every token and every comment. So if it helps to solve the problem, I wouldn't rule it out.

**Comment by scott-cotton** [comment]
@mdempsky 


> 
> I've been brainstorming lately about generalizing unified IR to a general serialization format for Go syntax+types and its possible benefits. (Ideally, that we only need to implement a single parser+typechecker that directly writes out a binary serialization; and then multiple, much simpler reader implementations that construct from that common serialization format either compiler IR, go/ast+go/types IR, or some hypothetical go2/* IR.)
> 
> In that world, I think it's practical to imagine keeping full-fidelity position information for every token and every comment. So if it helps to solve the problem, I wouldn't rule it out as too expensive.

Anything to share (online or DM) on this brainstorming?  I am interested.

**Comment by dave** [comment]
I've just released [dst v0.27.0](https://github.com/dave/dst) which adds compatibility with Go 1.18 generics. Feel free to have a play and let me know if it breaks anything. Many thanks to @hawkinsw for doing all the work on this upgrade!

**Comment by switchupcb** [comment]
Great! @dave

**Comment by dave** [comment]
This is awesome news. Thanks @adonovan ... as soon as I get in front of my laptop next week I'll have a good look through it. I'm very much looking forward to retiring the dst package!

**Comment by adonovan** [comment]
Thanks @dave, I appreciate the support but don't retire just yet: this is still just an early sketch of an approach.

**Comment by iamemilio** [comment]
This seems to be hanging. Is there some sort of expectation for when this may be ready to release? It would be nice to be able to depend on the standard library AST tools to manipulate syntax tree nodes rather than needing to roll your own or use third party tools. Are there are any stuck-points or questions that the community could help with?

In terms of preserving comment order for different field types, I think that the primary goal, from what I understand from the comments and my own use case, is to preserve the association between comments, spacing, and the lines of code (AST nodes) that they are associated with. With that in mind, I think your API that links node types to their positionally associated comments makes logical sense. As a user looking to read, manipulate or preserve the comments associated with a node, I would want to be aware of what comments are positioned above, inline with, or below it. Another possible approach is to make comments and spacing into their own distinct nodes and to add them directly to the tree.

**Comment by adonovan** [comment]
Not hanging, but it's the third priority on my to-do list; I hope to start on it this month. Glad to know this will be as useful to others as it will be to us.

**Comment by josharian** [comment]
> Glad to know this will be as useful to others as it will be to us.

It will! I hit https://github.com/golang/go/issues/22371 again just now.

**Comment by adonovan** [comment]
>  I hope to start on it this month.

Not even close, unfortunately. This will have to wait for go1.24.

**Comment by MattiasMartens** [comment]
This continues to be a sore spot in Go’s otherwise great tooling system. A parser whose output can’t re-generate the file it parsed?
This issue turned what could have been a twenty-minute task into one lasting several hours. It hadn’t occurred to me to imagine a widely-used library of a widely-used language prized for its pragmatic effectiveness would contain such a basic oversight.

**Comment by MattiasMartens** [comment]
> @MattiasMartens - agreed - did you try https://github.com/dave/dst ?

Great recommendation, thanks. To rebuild the code from tree is not needed for my case, and I can work around the problem without too much difficulty. I was just very surprised to encounter it.

**Comment by shahzadlone** [comment]
Looking forward for this, wondering if there is any notable remaining blockers that the community can help resolve to push the development on this further.

**Comment by griesemer** [comment]
@shahzadlone Thanks for your interest but I think this issue should be fixed by the Go team; it's very subtle and affects a lot of code. That said, there is some momentum in the team to get this fixed for 1.25 as part of a bigger push related to tooling.

**Comment by DanielMorsing** [comment]
I've done some experimentation on this issue over the past week and I want to present my findings, just to make sure that we don't duplicate efforts.

I think representing the comments directly in the AST nodes will be insufficient. Comments are really a property of the lexer. The only reason they exists as ranges in the AST is because they are bounded by the tokens that the lexer emits which then causes state transitions in the parser and produces ASTs. I think instead of representing them as attached to nodes in the grammar, we should represent them via state transitions. For example:

```go
if cond() /* something */ {
    // something else
    a = b
}
```
Imagine we represent comments as pairs of `inspector.Cursor` nodes instead. The first comment above would be represented by `[If.Cond, if.Body]`. The second would be represented as `[if.Body, Body.Assignstmt]`. The state transitions are arbitrary (within the grammar) but because the lexer must yield tokens to advance the state, the comment location is identified exactly.

Constructing this representation can be done without modifying the AST itself and can be done with no parser support. If the end user doesn't need to write out modified code, they can skip it entirely at no cost.

There are a couple of edge cases where the current AST representation isn't sufficient to uniquely identify where a token occurs. For example, `FieldList` and `CaseList` has no child with a position that covers the commas. This currently causes `gofmt` to move the tokens around (as noted [in Alan's CL](https://go-review.googlesource.com/c/go/+/429639/4/src/go/ast/precise.go#39)). I think there is a case for a minor change to the `go/ast` package to represent this.

Once we have comments identified in this scheme, printing a modified AST is a matter of checking whether a comment occurs between 2 constructions when emiting code. The biggest problem here will be adding to lists, but if we extend `Cursor` to calculate a hash tree with each event, matching should be fairly easy.

Anyway, these are some random thoughts. I'm currently working on a little skeleton to explore these ideas. I'll have a repo up when I've polished the code a bit.

EDIT: skeleton of the idea at https://github.com/DanielMorsing/commentStack

**Comment by adonovan** [comment]
Hi @DanielMorsing, sorry for the delay (vacation, Gophercon). Thanks for thinking about this problem. I feel like I'm missing something though: I think you are proposing that the caller of go/printer would supply an object that determines which comments are to be emitted as the printer works its way down and up the syntax tree. But the goal of this issue is to allow any "donor" ast.Node subtree to be grafted into a new "host" tree (either a different place in the same file, or a different file) so that printing of the modified "host" file generates comments within the "donor" portion just as they appeared in the original donor source file. That seems to me to imply that the comments are represented within the tree itself. By contrast the commentStack approach seems to require comments to be stored in some kind of external (e.g. per file) structure, not within the relevant subtree. That doesn't seem fundamentally different to the current ast.File.Comments table. But as I said, perhaps I am missing something.

FWIW, I think we could probably make the existing representation work somewhat better, for example by designing a set of operators for (e.g.) updating ast.File.Comments tables, shallow-copying ast.Files, renumbering Pos positions throughout the trees, constructing new token.Files with the new Poses, and so on. It's all rather fiddly and unsatisfactory though.

One thing we have learned is that, in the context of gopls, all ast.Files are "borrowed", since they are cached and shared in the parse cache, so no refactoring is allowed to modify them (unless it makes a deep copy first). But when you make a deep copy, you lose the ability to look up type information, since all the types.Info maps are keyed by ast.Node addresses. So any refactoring tools that want to splice and print trees need to forego type information in the modified copy. That's usually ok because the editing step is usually the last step in the tool.

Also, the desired output format for edits is not ast.Node trees, but textual edits: that's what LSP needs, for example, and it's the form used by the x/tools/go/analysis framework. So there's less need to splice trees than we once thought. The [inliner](https://pkg.go.dev/golang.org/x/tools/internal/refactor/inline) and [extract function](https://go.dev/gopls/features/transformation#refactorextract-extract-functionmethodvariable) operations are the main users of that approach today. Beyond inline and extract, the most common need we have for better comment handling is to compute the range of a syntax tree including its associated comments, for example, a statement and its preceding line comment and following end-of-line comment. This could be accomplished by just writing more library code using the existing data structures.

**Comment by DanielMorsing** [comment]
Hey @adonovan,

I will admit, the code in commentStack is a bit hard to follow. I got excited about what I thought was a good idea and knowing that tracking comments through AST manipulations was going to be a focus of development following your gophercon talk, I wanted to get something written down fast for discussion. It shows that it was all done in a bit of a hurry.

I took my lead from [CL429639](https://go-review.googlesource.com/c/go/+/429639), thinking that grafting would be the main way that ASTs would be modified. Since I wrote the last comment, I've done some work on how to anchor comments between `ast.Node`s under arbitrary AST tree edits, but I suspect a lot of this work will not be very relevant if the main method we see in the future is textual edits that already track of the various offsets within the file.

There is one idea that I think is helpful from commentStack. For every comment, I calculate the last non-comment token before it and the next non-comment token after it by rerunning the lexer over a file. From these, we can figure out what "bounds" a comment, even in cases where we don't store position information (for example: the `.` in `SelectorExpr`). When moving expressions, we can then expand the ranges to include the comments that might be associated with it.

This obviously won't work for gopls as is, because it gets its ASTs from the parse cache, but it's a small change to the standard library to add this information. We could also do it when we do the parsing, storing it in the cache but outside the AST.

PS. before I realised the implications of `TextEdits`, I wrote a comment for this issue that explains my approach a bit better. If there's any interest in reading it, I've left it in the [README](https://github.com/DanielMorsing/tokenorder/blob/main/README.md) of commentStack.

**Comment by adonovan** [comment]
Aside: One problem that is not solved by the approach in [CL 429639](https://go.dev/cl/429639) is that of vertical space. We have had many reports that the code transformation done by the [inliner](https://pkg.go.dev/golang.org/x/tools/gopls/internal/analysis/gofix) is often undesirable because, even ignoring the issue of comment preservation, it discards vertical space. For example, an expression of the form

```go
T{
  k: v,
  k: v,
}
```

might be transformed to `T{k: v, k: v}`. If there are many elements in the literal, the resulting line can be hundreds of characters long.

Ideally the syntax tree would record the positions of newlines (just as it should comments) and provide this information to the formatter. I don't have any good ideas for how to do this yet, but I think it is not enough to solve only the comments problem.

(I would love to redesign go/ast from scratch so that it faithfully records the position of every comment and token, even in the case of parse errors. I have lots of ideas for how to do this (and how it could provide the inspector.Cursor functionality directly). Unfortunately I don't see any way that this can be achieved without undermining all the benefits of having exactly one standard AST representation. For example, the go/types package would need to be forked to work on another representation---just as it was to support the compiler's [syntax.Node](https://pkg.go.dev/cmd/compile/internal/syntax#Node), which continues to be a significant burden; and the Go analysis framework exposes its dependency on go/ast and go/types. Whatever solution we come up with must work with go/ast.)

**Comment by mateusz834** [comment]
I have also been thinking about this (for fun), and i think you can get pretty far with the current free-floating comments. I believe that it is doable with this kind of API:

```go
type File struct {
	File *ast.File
	Fset *token.FileSet
}

// Pos returns a new position on the current line.
func (p *PosBuilder) Pos() token.Pos {
}

// Newline returns a new position on a new line (col == 1).
func (p *PosBuilder) Newline() token.Pos {
}

// WithPosBuilder executes build with a PosBuilder, that returns positions between stard and end (returnted by
// the between func). If there are no enough positons available, then the f.File is expanded (i.e. new
// *token.File is created and each positon is rewritten so thare are more positions in between [token.Pos]s
// in the AST).
//
// between must always be relative (e.g. return f.Decls[0].End(), f.FileEnd) and reference the positions in the file, so
// that after the rewrite (expansion of the AST) new positions are returned.
//
// build/between might be called multiple time until it stabilizes.
func WithPosBuilder(f *File, between func() (start, end token.Pos), build func(b *PosBuilder)) {
}

```

Which could be used as:

```go
var body ast.BlockStmt
var c ast.Comment
var c2 ast.Comment

WithPosBuilder(&file, func() (start token.Pos, end token.Pos) {
	return f.Decls[0].End(), f.FileEnd
}, func(b *PosBuilder) {
	body.Lbrace = b.Newline()
	c.Slash = b.Newline()
	c.Text = "// An assignment"
	body.List = []ast.Stmt{
		&ast.AssignStmt{
			Lhs: []ast.Expr{
				&ast.Ident{
					NamePos: b.Newline(),
					Name:    "foo",
				},
			},
			TokPos: b.Pos(),
			Tok:    token.DEFINE,
			Rhs: []ast.Expr{
				&ast.Ident{
					NamePos: b.Pos(),
					Name:    "foo",
				},
			},
		},
	}
	c2.Slash = b.Pos()
	c2.Text = "// this is an assigment"
	body.Rbrace = b.Newline()
})

// insert new nodes into the AST
```

This approach might be slow (since it has to walk and rewrite the entire AST, can be done once, like expand by 1k positions beforehand), but it seems to give a better control of comments in the AST. I don't have yet a working solution for this, but it does not seem that complicated. This approach works without any change to go/ast, maybe some slight changes in go/printer might be necessary (such that the expansion process does not change the results of the printer)

**Comment by willfaught** [comment]
I tried to solve this problem years ago, and played around with a few ideas. My first idea was to have every node type contain the whitespace, line, and comment syntax ("padding") before and after it, and allow padding as decls, specs, and statements. You can accurately represent a lot of syntax that way, but there are three problems with it: (1) the AST doesn't explicitly represent some parts of the grammar, such as commas and semicolons, so you can't track and preserve padding in relation to them; (2) there is at least one place in the grammar where two tokens are adjacent within the same node (GenDecl.TokPos and GenDecl.Lparen, e.g. `var (`), so the padding between the tokens wouldn't be represented; and (3) file positions (possibly all of them) have to be manually updated as you add, move, resize, or remove nodes to keep them consistent.

My second idea was for a node to be a list of tokens straight out of the lexer token stream, including padding tokens (including new whitespace and line tokens). Instead of node fields, there would be node methods that would scan the token list to find the corresponding token or node (or compute the doc, say). The padding tokens before and after the node would be at the beginning and end of the token list respectively. This seems to solve the three problems above. However, the methods have to do work every time they're called, token lists can be missing vital elements without validation, and token lists could be long and cumbersome to work with.

I think of the current AST design as a lightweight, read-only representation of a file. It tracks positions and padding elsewhere in FileSet and File.Comments to stay lean. It's geared around reading syntax and getting the position of nodes. However, for writing syntax, knowing the position of nodes isn't important because it's implicit in their place in the node structure. The printer will naturally print nodes at their "position" as it visits them. If a position is actually needed, it's better to compute it dynamically on demand, rather than compute the position for every node up front and cache it in the node, then have to recompute those positions every time an edit affects them. Likewise, if you want to edit whitespace, lines, or comments, it's better to do it in place, in the node structure, rather than somewhere else like FileSet or File.Comments. Putting padding inside nodes makes nodes self-contained, and thus portable, which facilitates editing. Such a portable design is (I think inevitably) heavyweight. Perhaps the best we can do is enable conversion between lightweight and heavyweight AST designs.
